### PR TITLE
feat(update-server): Add system update support for OT3 + update file validation

### PR DESCRIPTION
**Pull request by vegano1**
[RCORE-320](https://opentrons.atlassian.net/browse/RCORE-320)

# Overview
The update-server runs on both the OT2 and OT3 however they have different implementations because the underlying system is different. In order to support this we have an interface that both the OT2 and OT3 use to be able to communicate this updated information both in its internal API and to the clients (electron) apps as well. This adds new "systemUpdate" capabilities for the openembedded class (OT3) and buildroot class (OT2).

# Changelog
- Added systemUpdate capabilities to buildroot/openemdedded health_endpoints
- Added UPDATE_PKG_BR and UPDATE_PKG_OE to know which file upload fields are accepted by the different updaters
- Added verification of the update using the new robot_model flag in the VERSION.json file so we only install updates whose robot_model matches the running instance.
- Added InvalidPKGName to tell clients if the field name they sent is invalid
- Added InvalidRobotModel to tell clients that the robot_model in the VERSION.json file is invalid
- Renamed openembedded updater to OT3UpdateActions for consistency
- Use context manager when opening hash file so we don't leave behind open file descriptors
- Moved load_version_file to common so it is shared between the updaters

# Review requests
- [x] Make sure OT2 reports buildroot/systemUpdate capabilities in **server/update/health** endpoint
- [x] Make sure OT2 accepts both 'ot2-system.zip' and 'system-update.zip' file fields when uploading update via **/server/update/{token}/file endpoint**
- [x] Make sure OT3 reports systemUpdate capability in **server/update/health** endpoint
- [x] Make sure OT3 accepts 'system-update.zip' file field when uploading update via **/server/update/{token}/file** endpoint
- [x] Make sure that update files are only processed by the correct robot type based on the 'robot_model' flag in the VERSION.json file. So 'ot2-system.zip' files are only handled by the OT2 and 'system-update.zip' files are handled by the OT3.

Note: The 'robot_model' is a new flag added to the VERSION.json file during the build which represents the robot model this update was created for. It will be added as part of this effort and can be found here for [Buildroot](https://github.com/Opentrons/buildroot/pull/185) and [Openembedded](https://github.com/Opentrons/oe-core/pull/27)

# Risk assessment
Low, changes are primarily for OT3 updates but also minimally affect the OT2 updates for backward compatibility.

# Todo
- [x] add test code coverage, will add in separate pr.

### CODE REVIEW

File: update-server/otupdate/common/file_actions.py [hunk 1]
```diff
@@ -220,3 +245,14 @@ def verify_signature(message_path: str, sigfile_path: str, cert_path: str) -> No
...
+
+
+def load_version_file(version_file: str) -> Mapping[str, str]:
+    version = {}
+    LOG.debug(f"Loading version file {version_file}")
+    try:
+        with open(version_file, "r") as fh:
+            version = json.load(fh)
+    except Exception:
+        logging.exception(f"Could not load version file - {version_file}")
```
Comment (sfoster1) [inline update-server/otupdate/common/file_actions.py]:
```suggestion
        LOG.exception(f"Could not load version file - {version_file}")
```

**Review by sfoster1 — COMMENTED** [review]
Code looks good to me, but haven't tested yet so won't approve (and we want the extra tests as you noted in the PR description of course)

File: update-server/otupdate/common/update.py [hunk 2]
```diff
@@ -172,7 +176,7 @@ async def file_upload(request: web.Request, session: UpdateSession) -> web.Respo
     """Serves /update/:session/file
 
     Requires multipart (encoding doesn't matter) with a file field in the
-    body called 'ot2-system.zip'.
+    body called 'system-update.zip' or 'ot2-system.zip'.
```
Comment (sfoster1) [inline update-server/otupdate/common/update.py]: [nit] we should just call out `system-update.zip` here, with maybe a note that ot2s will also support `ot2-system.zip`

File: update-server/otupdate/common/update.py [hunk 3]
```diff
@@ -184,13 +188,11 @@ async def file_upload(request: web.Request, session: UpdateSession) -> web.Respo
...
     async for part in reader:
-        # TODO (al, 2022-04-18): This check should not be here. All ot2 and
-        #  ot3 disambiguation should happen in update actions.
-        if part.name != "ot2-system.zip" and part.name != UPDATE_PKG:
+        if part.name not in VALID_UPDATE_PKG:
             LOG.info(f"Unknown field name {part.name} in file_upload, ignoring")
             await part.release()
         else:
-            LOG.info(f"Writing {part.name}")
+            LOG.info(f"Downloading {part.name}")
```
Comment (sfoster1) [inline update-server/otupdate/common/update.py]: [nit] i really don't like the use of "downloading" here - imo "downloading" means a server->client data transfer, and this is a client->server.

also, it really is just writing this data. let's remove some of these log changes.

File: update-server/tests/common/test_update.py [hunk 4]
```diff
@@ -197,6 +198,7 @@ async def test_update_happypath(
         resp = await test_cli[0].get(session_endpoint(update_session, "status"))
         assert resp.status == 200
         body = await resp.json()
+        print("YOO", body)
```
Comment (sfoster1) [inline update-server/tests/common/test_update.py]: debug print

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Couple nits about logging, and also a debug print; functionality looks good though!

**Review by shlokamin — APPROVED** [review]
Tested OT-2 + OT-3 system updates and this works like a charm!

Only thing to note is that my capabilities map for the OT-2 looks like this: 

```
"capabilities": {"systemUpdate": "/server/update/begin", "buildrootUpdate": "/server/update/begin", "restart": "/server/restart"}

```

which is slightly different than this: 

> - [ ] Make sure OT2 reports buildroot/systemUpdate capabilities in **server/update/health** endpoint

**Comment by vegano1** [comment]
> Tested OT-2 + OT-3 system updates and this works like a charm!
> 
> Only thing to note is that my capabilities map for the OT-2 looks like this:
> 
> ```
> "capabilities": {"systemUpdate": "/server/update/begin", "buildrootUpdate": "/server/update/begin", "restart": "/server/restart"}
> ```
> 
> which is slightly different than this:
> 
> > * [ ]  Make sure OT2 reports buildroot/systemUpdate capabilities in **server/update/health** endpoint

yep, forgot to update the comment but that looks good!

**Review by sfoster1 — APPROVED** [review]
Looks awesome to me!

**Pull request merged**